### ISSUE TITLE
🐞[BUG]: When using @select, If the passed method throws an Exception, the subscription stops and do not emits any values after that.

### ISSUE THREAD

**Issue by shivamd20** [ngxs/store#995]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ x ] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => https://github.com/ngxs/store/blob/master/CONTRIBUTING.md
[ ] Other... Please describe:
</code></pre>

## Current behavior

When using @Select, If the passed method throws an Exception the subscription stops and do not emits any values after that.

## Expected behavior
Selector should ignore the particular  case where the passed method throws an exception and it should print the error message  on browser console.
We should get subsequent updates even when we get error on the selector method.

## Minimal reproduction of the problem with instructions

> Lets say this is our data model
```  
export class AuthStateModel {
    profile: Object;
}
  ```
> and this is our state 

```  

@State<AuthStateModel>({
    name: 'auth',
    defaults: {
        profile: {},
    }
})
export class AuthState {


    constructor(private authService: AuthService) { }

    @Action(ClearAuth)
    clearAuth(ctx: StateContext<AuthStateModel>) {
        ctx.setState({
            profile: undefined
        });
    }

    @Action(SaveAuthData)
    saveAuthData(ctx: StateContext<AuthStateModel>, action: SaveAuthData) {
        const state = ctx.getState();
        const nextState = produce(state, (draftState) => Object.assign(draftState, { ...action.payload }));
        ctx.setState(nextState);
    }

// more actions 
}
```
> Now if we write a selector like this

` @Select(state =>  state.auth.profile.picture )  `
And then if we dispatch two actions,   @Action(ClearAuth) and @Action(SaveAuthData), We will not get the data  from second action because the first action will make the profile object undefined and when ngxs calls select's passed method, It will throw ` cannot read property length of undefined ` exception.

## Environment

<pre><code>
Libs:
- @angular/core version: 7.0.4
- @ngxs/store version: 3.3.2

Browser:
- [x] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX

</code></pre>

**Comment by splincode** [comment]
Please add stackblitz for reproduce

**Comment by shivamd20** [comment]
https://stackblitz.com/edit/ngxs-simple-2c6fdk

**Comment by splincode** [comment]
![image](https://user-images.githubusercontent.com/12021443/61562899-069a8500-aa7b-11e9-95be-b11683a3142b.png)

**Comment by splincode** [comment]
You do not need to throw an exception up!
https://github.com/ngxs/store/pull/1164/files

**Comment by markwhitfeld** [comment]
I think that this is a bug, reopening.
The `@Select` observable should definitely not close on error.
if you use the store directly then I would expect the user to do the error handling (if `supressErrors` is false) on the observable returned by `this.store.select(...)`.

**Comment by shivamd20** [comment]
> You do not need to throw an exception up!
> https://github.com/ngxs/store/pull/1164/files

Yes, I know. The example was to show that the subscription was getting stopped after an exception has been thrown. Surrounding it with try-catch is a workaround for it. The Subscription should not stop even if there is an exception.

**Comment by splincode** [comment]
@shivamd20 @markwhitfeld  After some research, I still understand that this is the normal behavior of RxJS. An explanation of why this matters can be found from Rx design guidelines:

> The single message indicating that an observable sequence has finished ensures that consumers of the observable sequence can deterministically establish that it is safe to perform cleanup operations.
A single failure further ensures that abort semantics can be maintained for operators that work on multiple observable sequences.

Therefore, if you want everything to work correctly, you need to understand where the error occurs, use `suppressErrors`

```ts
NgxsModule.forRoot([CountState], { 
 selectorOptions: {
    suppressErrors: false
 }
})
export class AppModule {}
```

One of the problems is that if you break the code execution flow, then naturally, RxJS will unsubscribe, since if you want to re-subscribe again, in this case, an error will occur cyclically.

#### Why does Rxjs unsubscribe on error?

An explanation of why this matters can be found from Rx design guidelines:

The single message indicating that an observable sequence has finished ensures that consumers of the observable sequence can deterministically establish that it is safe to perform cleanup operations.

A single failure further ensures that abort semantics can be maintained for operators that work on multiple observable sequences.

In short, if you want your observers to keep listening to the store after a error has occurred, do not deliver that error to the store, but rather handle it in some other way (e.g. use catch, retry or deliver the error to a dedicated subject).

```ts
this.store.select(CountState).pipe(
  retryWhen(errors =>
    errors.pipe(
      // log error message
      tap(val => console.log(`Value ${val} was too high!`)),
      // restart in 6 seconds
      delayWhen(val => timer(val * 1000))
    )
  )
);
```
